File: src/main/java/org/jcodec/containers/mp4/boxes/TrunBox.java
package org.jcodec.containers.mp4.boxes;

import java.nio.ByteBuffer;
import java.util.List;

//@formatter:off
/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * Track fragment run
 * 
 * To crate new box:
 * 
 * <pre>
 * 
 * Box box = TrunBox
 *         .create(2)
 *         .dataOffset(20)
 *         .sampleCompositionOffset(new int[] { 11, 12 })
 *         .sampleDuration(new int[] { 15, 16 })
 *         .sampleFlags(new int[] { 100, 200 })
 *         .sampleSize(new int[] { 30, 40 })
 *         .create();
 * 
 * </pre>
 * 
 * @author The JCodec project
 * 
 */
//@formatter:on
public class TrunBox extends FullBox {
    // @formatter:off
    private static final int DATA_OFFSET_AVAILABLE = 0x000001;
    private static final int FIRST_SAMPLE_FLAGS_AVAILABLE = 0x000004;
    private static final int SAMPLE_DURATION_AVAILABLE = 0x000100;
    private static final int SAMPLE_SIZE_AVAILABLE = 0x000200;
    private static final int SAMPLE_FLAGS_AVAILABLE = 0x000400;
    private static final int SAMPLE_COMPOSITION_OFFSET_AVAILABLE = 0x000800;
    // @formatter:on

    private int sampleCount;
    private int dataOffset;
    private int firstSampleFlags;
    private int[] sampleDuration;
    private int[] sampleSize;
    private int[] sampleFlags;
    private int[] sampleCompositionOffset;

    public static String fourcc() {
        return "trun";
    }

    public void setDataOffset(int dataOffset) {
        this.dataOffset = dataOffset;
    }

    public static Factory create(int sampleCount) {
        return new Factory(TrunBox.createTrunBox1(sampleCount));
    }

    public static Factory copy(TrunBox other) {
        TrunBox box = TrunBox
                .createTrunBox2(other.sampleCount, other.dataOffset, other.firstSampleFlags, other.sampleDuration, other.sampleSize, other.sampleFlags, other.sampleCompositionOffset);
        box.setFlags(other.getFlags());
        box.setVersion(other.getVersion());
        return new Factory(box);
    }

    public TrunBox(Header header) {
        super(header);
    }

    public static TrunBox createTrunBox1(int sampleCount) {
        TrunBox trun = new TrunBox(new Header(fourcc()));
        trun.sampleCount = sampleCount;
        return trun;
    }

    public static TrunBox createTrunBox2(int sampleCount, int dataOffset, int firstSampleFlags, int[] sampleDuration,
            int[] sampleSize, int[] sampleFlags, int[] sampleCompositionOffset) {
        TrunBox trun = new TrunBox(new Header(fourcc()));
        trun.sampleCount = sampleCount;
        trun.dataOffset = dataOffset;
        trun.firstSampleFlags = firstSampleFlags;
        trun.sampleDuration = sampleDuration;
        trun.sampleSize = sampleSize;
        trun.sampleFlags = sampleFlags;
        trun.sampleCompositionOffset = sampleCompositionOffset;
        return trun;
    }

    public static class Factory {

        private TrunBox box;

        protected Factory(TrunBox box) {
            this.box = box;
        }

        public Factory dataOffset(long dataOffset) {
            box.flags |= DATA_OFFSET_AVAILABLE;
            box.dataOffset = (int) dataOffset;
            return this;
        }

        public Factory firstSampleFlags(int firstSampleFlags) {
            if (box.isSampleFlagsAvailable())
                throw new IllegalStateException("Sample flags already set on this object");
            box.flags |= FIRST_SAMPLE_FLAGS_AVAILABLE;
            box.firstSampleFlags = firstSampleFlags;
            return this;
        }

        public Factory sampleDuration(int[] sampleDuration) {
            if (sampleDuration.length != box.sampleCount)
                throw new IllegalArgumentException("Argument array length not equal to sampleCount");
            box.flags |= SAMPLE_DURATION_AVAILABLE;
            box.sampleDuration = sampleDuration;
            return this;
        }

        public Factory sampleSize(int[] sampleSize) {
            if (sampleSize.length != box.sampleCount)
                throw new IllegalArgumentException("Argument array length not equal to sampleCount");
            box.flags |= SAMPLE_SIZE_AVAILABLE;
            box.sampleSize = sampleSize;
            return this;
        }

        public Factory sampleFlags(int[] sampleFlags) {
            if (sampleFlags.length != box.sampleCount)
                throw new IllegalArgumentException("Argument array length not equal to sampleCount");
            if (box.isFirstSampleFlagsAvailable())
                throw new IllegalStateException("First sample flags already set on this object");
            box.flags |= SAMPLE_FLAGS_AVAILABLE;
            box.sampleFlags = sampleFlags;
            return this;
        }

        public Factory sampleCompositionOffset(int[] sampleCompositionOffset) {
            if (sampleCompositionOffset.length != box.sampleCount)
                throw new IllegalArgumentException("Argument array length not equal to sampleCount");
            box.flags |= SAMPLE_COMPOSITION_OFFSET_AVAILABLE;
            box.sampleCompositionOffset = sampleCompositionOffset;
            return this;
        }

        public TrunBox create() {
            try {
                return box;
            } finally {
                box = null;
            }
        }
    }

    public long getSampleCount() {
        return sampleCount & 0xffffffffL;
    }

    public int getDataOffset() {
        return dataOffset;
    }

    public int getFirstSampleFlags() {
        return firstSampleFlags;
    }

    public int[] getSampleDurations() {
        return sampleDuration;
    }

    public int[] getSampleSizes() {
        return sampleSize;
    }

    public int[] getSamplesFlags() {
        return sampleFlags;
    }

    public int[] getSampleCompositionOffsets() {
        return sampleCompositionOffset;
    }

    public long getSampleDuration(int i) {
        return sampleDuration[i] & 0xffffffffL;
    }

    public long getSampleSize(int i) {
        return sampleSize[i] & 0xffffffffL;
    }

    public int getSampleFlags(int i) {
        return sampleFlags[i];
    }

    public long getSampleCompositionOffset(int i) {
        return sampleCompositionOffset[i] & 0xffffffffL;
    }

    public boolean isDataOffsetAvailable() {
        return (flags & DATA_OFFSET_AVAILABLE) != 0;
    }

    public boolean isSampleCompositionOffsetAvailable() {
        return (flags & SAMPLE_COMPOSITION_OFFSET_AVAILABLE) != 0;
    }

    public boolean isSampleFlagsAvailable() {
        return (flags & SAMPLE_FLAGS_AVAILABLE) != 0;
    }

    public boolean isSampleSizeAvailable() {
        return (flags & SAMPLE_SIZE_AVAILABLE) != 0;
    }

    public boolean isSampleDurationAvailable() {
        return (flags & SAMPLE_DURATION_AVAILABLE) != 0;
    }

    public boolean isFirstSampleFlagsAvailable() {
        return (flags & FIRST_SAMPLE_FLAGS_AVAILABLE) != 0;
    }

    public static int flagsGetSampleDependsOn(int flags) {
        return (flags >> 6) & 0x3;
    }

    public static int flagsGetSampleIsDependedOn(int flags) {
        return (flags >> 8) & 0x3;
    }

    public static int flagsGetSampleHasRedundancy(int flags) {
        return (flags >> 10) & 0x3;
    }

    public static int flagsGetSamplePaddingValue(int flags) {
        return (flags >> 12) & 0x7;
    }

    public static int flagsGetSampleIsDifferentSample(int flags) {
        return (flags >> 15) & 0x1;
    }

    public static int flagsGetSampleDegradationPriority(int flags) {
        return (flags >> 16) & 0xffff;
    }

    public static TrunBox createTrunBox() {
        return new TrunBox(new Header(fourcc()));
    }

    @Override
    public void parse(ByteBuffer input) {
        super.parse(input);

        if (isSampleFlagsAvailable() && isFirstSampleFlagsAvailable())
            throw new RuntimeException("Broken stream");

        sampleCount = input.getInt();
        if (isDataOffsetAvailable())
            dataOffset = input.getInt();
        if (isFirstSampleFlagsAvailable())
            firstSampleFlags = input.getInt();
        if (isSampleDurationAvailable())
            sampleDuration = new int[sampleCount];
        if (isSampleSizeAvailable())
            sampleSize = new int[sampleCount];
        if (isSampleFlagsAvailable())
            sampleFlags = new int[sampleCount];
        if (isSampleCompositionOffsetAvailable())
            sampleCompositionOffset = new int[sampleCount];

        for (int i = 0; i < sampleCount; i++) {
            if (isSampleDurationAvailable())
                sampleDuration[i] = input.getInt();
            if (isSampleSizeAvailable())
                sampleSize[i] = input.getInt();
            if (isSampleFlagsAvailable())
                sampleFlags[i] = input.getInt();
            if (isSampleCompositionOffsetAvailable())
                sampleCompositionOffset[i] = input.getInt();
        }
    }

    @Override
    public void doWrite(ByteBuffer out) {
        super.doWrite(out);
        out.putInt(sampleCount);
        if (isDataOffsetAvailable())
            out.putInt(dataOffset);
        if (isFirstSampleFlagsAvailable())
            out.putInt(firstSampleFlags);

        for (int i = 0; i < sampleCount; i++) {
            if (isSampleDurationAvailable())
                out.putInt(sampleDuration[i]);
            if (isSampleSizeAvailable())
                out.putInt(sampleSize[i]);
            if (isSampleFlagsAvailable())
                out.putInt(sampleFlags[i]);
            if (isSampleCompositionOffsetAvailable())
                out.putInt(sampleCompositionOffset[i]);
        }
    }

    protected void getModelFields(List<String> model) {
        model.add("sampleCount");
        if (isDataOffsetAvailable())
            model.add("dataOffset");

        if (isFirstSampleFlagsAvailable())
            model.add("firstSampleFlags");

        if (isSampleDurationAvailable())
            model.add("sampleDuration");

        if (isSampleSizeAvailable())
            model.add("sampleSize");

        if (isSampleFlagsAvailable())
            model.add("sampleFlags");

        if (isSampleCompositionOffsetAvailable())
            model.add("sampleCompositionOffset");
    }
}

File: src/main/java/org/jcodec/containers/mp4/boxes/TrunBox.kt
package org.jcodec.containers.mp4.boxes

import org.jcodec.platform.Platform
import java.nio.ByteBuffer

//@formatter:off
/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * Track fragment run
 *
 * To crate new box:
 *
 * <pre>
 *
 * Box box = TrunBox
 * .create(2)
 * .dataOffset(20)
 * .sampleCompositionOffset(new int[] { 11, 12 })
 * .sampleDuration(new int[] { 15, 16 })
 * .sampleFlags(new int[] { 100, 200 })
 * .sampleSize(new int[] { 30, 40 })
 * .create();
 *
</pre> *
 *
 * @author The JCodec project
 */
//@formatter:on
class TrunBox(header: Header) : FullBox(header) {
    // @formatter:on
    private var sampleCount = 0
    private var dataOffset: Long = 0
    var firstSampleFlags = 0
        private set
    var sampleDurations: IntArray = IntArray(0)
        private set
    var sampleSizes: IntArray = IntArray(0)
        private set
    var samplesFlags: IntArray = IntArray(0)
        private set
    var sampleCompositionOffsets: IntArray = IntArray(0)
        private set

    fun setDataOffset(dataOffset: Long) {
        this.dataOffset = dataOffset
    }

    class Factory(private var box: TrunBox?) {
        fun dataOffset(dataOffset: Long): Factory {
            box!!.flags = box!!.flags or DATA_OFFSET_AVAILABLE
            box!!.dataOffset = (dataOffset.toInt()).toLong()
            return this
        }

        fun firstSampleFlags(firstSampleFlags: Int): Factory {
            check(!box!!.isSampleFlagsAvailable) { "Sample flags already set on this object" }
            box!!.flags = box!!.flags or FIRST_SAMPLE_FLAGS_AVAILABLE
            box!!.firstSampleFlags = firstSampleFlags
            return this
        }

        fun sampleDuration(sampleDuration: IntArray): Factory {
            require(sampleDuration.size == box!!.sampleCount) { "Argument array length not equal to sampleCount" }
            box!!.flags = box!!.flags or SAMPLE_DURATION_AVAILABLE
            box!!.sampleDurations = sampleDuration
            return this
        }

        fun sampleSize(sampleSize: IntArray): Factory {
            require(sampleSize.size == box!!.sampleCount) { "Argument array length not equal to sampleCount" }
            box!!.flags = box!!.flags or SAMPLE_SIZE_AVAILABLE
            box!!.sampleSizes = sampleSize
            return this
        }

        fun sampleFlags(sampleFlags: IntArray): Factory {
            require(sampleFlags.size == box!!.sampleCount) { "Argument array length not equal to sampleCount" }
            check(!box!!.isFirstSampleFlagsAvailable) { "First sample flags already set on this object" }
            box!!.flags = box!!.flags or SAMPLE_FLAGS_AVAILABLE
            box!!.samplesFlags = sampleFlags
            return this
        }

        fun sampleCompositionOffset(sampleCompositionOffset: IntArray): Factory {
            require(sampleCompositionOffset.size == box!!.sampleCount) { "Argument array length not equal to sampleCount" }
            box!!.flags = box!!.flags or SAMPLE_COMPOSITION_OFFSET_AVAILABLE
            box!!.sampleCompositionOffsets = sampleCompositionOffset
            return this
        }

        fun create(): TrunBox? {
            return try {
                box
            } finally {
                box = null
            }
        }

    }

    fun getSampleCount(): Long {
        return Platform.unsignedInt(sampleCount)
    }

    fun getDataOffset(): Long {
        return dataOffset
    }

    fun getSampleDuration(i: Int): Long {
        return Platform.unsignedInt(sampleDurations[i])
    }

    fun getSampleSize(i: Int): Long {
        return Platform.unsignedInt(sampleSizes[i])
    }

    fun getSampleFlags(i: Int): Int {
        return samplesFlags[i]
    }

    fun getSampleCompositionOffset(i: Int): Long {
        return Platform.unsignedInt(sampleCompositionOffsets[i])
    }

    val isDataOffsetAvailable: Boolean
        get() = flags and DATA_OFFSET_AVAILABLE != 0

    val isSampleCompositionOffsetAvailable: Boolean
        get() = flags and SAMPLE_COMPOSITION_OFFSET_AVAILABLE != 0

    val isSampleFlagsAvailable: Boolean
        get() = flags and SAMPLE_FLAGS_AVAILABLE != 0

    val isSampleSizeAvailable: Boolean
        get() = flags and SAMPLE_SIZE_AVAILABLE != 0

    val isSampleDurationAvailable: Boolean
        get() = flags and SAMPLE_DURATION_AVAILABLE != 0

    val isFirstSampleFlagsAvailable: Boolean
        get() = flags and FIRST_SAMPLE_FLAGS_AVAILABLE != 0

    override fun parse(input: ByteBuffer) {
        super.parse(input)
        if (isSampleFlagsAvailable && isFirstSampleFlagsAvailable) throw RuntimeException("Broken stream")
        sampleCount = input.int
        if (isDataOffsetAvailable) dataOffset = (input.int.toLong() and 0xffffffffL)
        if (isFirstSampleFlagsAvailable) firstSampleFlags = input.int
        if (isSampleDurationAvailable) sampleDurations = IntArray(sampleCount)
        if (isSampleSizeAvailable) sampleSizes = IntArray(sampleCount)
        if (isSampleFlagsAvailable) samplesFlags = IntArray(sampleCount)
        if (isSampleCompositionOffsetAvailable) sampleCompositionOffsets = IntArray(sampleCount)
        for (i in 0 until sampleCount) {
            if (isSampleDurationAvailable) sampleDurations[i] = input.int
            if (isSampleSizeAvailable) sampleSizes[i] = input.int
            if (isSampleFlagsAvailable) samplesFlags[i] = input.int
            if (isSampleCompositionOffsetAvailable) sampleCompositionOffsets[i] = input.int
        }
    }

    public override fun doWrite(out: ByteBuffer) {
        super.doWrite(out)
        out.putInt(sampleCount)
        if (isDataOffsetAvailable) out.putInt(dataOffset.toInt())
        if (isFirstSampleFlagsAvailable) out.putInt(firstSampleFlags)
        for (i in 0 until sampleCount) {
            if (isSampleDurationAvailable) out.putInt(sampleDurations[i])
            if (isSampleSizeAvailable) out.putInt(sampleSizes[i])
            if (isSampleFlagsAvailable) out.putInt(samplesFlags[i])
            if (isSampleCompositionOffsetAvailable) out.putInt(sampleCompositionOffsets[i])
        }
    }

    override fun estimateSize(): Int {
        return 24 + sampleCount * 16
    }

    companion object {
        // @formatter:off
        private const val DATA_OFFSET_AVAILABLE = 0x000001
        private const val FIRST_SAMPLE_FLAGS_AVAILABLE = 0x000004
        private const val SAMPLE_DURATION_AVAILABLE = 0x000100
        private const val SAMPLE_SIZE_AVAILABLE = 0x000200
        private const val SAMPLE_FLAGS_AVAILABLE = 0x000400
        private const val SAMPLE_COMPOSITION_OFFSET_AVAILABLE = 0x000800
        @JvmStatic
        fun fourcc(): String {
            return "trun"
        }

        @JvmStatic
        fun create(sampleCount: Int): Factory {
            return Factory(createTrunBox1(sampleCount))
        }

        fun copy(other: TrunBox): Factory {
            val box = createTrunBox2(other.sampleCount, other.dataOffset, other.firstSampleFlags, other.sampleDurations, other.sampleSizes, other.samplesFlags, other.sampleCompositionOffsets)
            box.flags = other.flags
            box.version = other.version
            return Factory(box)
        }

        fun createTrunBox1(sampleCount: Int): TrunBox {
            val trun = TrunBox(Header(fourcc()))
            trun.sampleCount = sampleCount
            return trun
        }

        fun createTrunBox2(sampleCount: Int, dataOffset: Long, firstSampleFlags: Int, sampleDuration: IntArray,
                           sampleSize: IntArray, sampleFlags: IntArray, sampleCompositionOffset: IntArray): TrunBox {
            val trun = TrunBox(Header(fourcc()))
            trun.sampleCount = sampleCount
            trun.dataOffset = dataOffset
            trun.firstSampleFlags = firstSampleFlags
            trun.sampleDurations = sampleDuration
            trun.sampleSizes = sampleSize
            trun.samplesFlags = sampleFlags
            trun.sampleCompositionOffsets = sampleCompositionOffset
            return trun
        }

        fun flagsGetSampleDependsOn(flags: Int): Int {
            return flags shr 6 and 0x3
        }

        fun flagsGetSampleIsDependedOn(flags: Int): Int {
            return flags shr 8 and 0x3
        }

        fun flagsGetSampleHasRedundancy(flags: Int): Int {
            return flags shr 10 and 0x3
        }

        fun flagsGetSamplePaddingValue(flags: Int): Int {
            return flags shr 12 and 0x7
        }

        fun flagsGetSampleIsDifferentSample(flags: Int): Int {
            return flags shr 15 and 0x1
        }

        fun flagsGetSampleDegradationPriority(flags: Int): Int {
            return flags shr 16 and 0xffff
        }

        @JvmStatic
        fun createTrunBox(): TrunBox {
            return TrunBox(Header(fourcc()))
        }
    }
}